### PR TITLE
Android Nimbus JS Distribution

### PR DESCRIPTION
**Pull request by paultiarks**
This does many things. A new module `nimbusjs` has been created in the Android project. In the gradle file for this module is executing a new bash script which builds nimbus, and copies the resulting nimbus.js into the resources of the new module. A utility class has been created that finds the `<head>` tag of an `InputStream` and injects nimbus.js. This will allow consumers to intercept page requests in their `WebViewClient` and use this utility function to inject nimbus.

All feedback is more than welcome, this is my first foray into the Android world and I expect there will need to be changes.

### CODE REVIEW

File: platforms/android/nimbusjs/src/androidTest/java/NimbusJSUtilsTests.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        var containsNimbus = resultString?.contains("nimbus")
+        assertEquals(true, containsNimbus)
+    }
+
+    private fun readAssetStream(stream: InputStream): String? {
+        try {
+            val bufferSize = 1024
+            val buffer = CharArray(bufferSize)
+            val out = StringBuilder()
+            val `in`: Reader = InputStreamReader(stream, "UTF-8")
```
Comment (jamie-houston) [inline platforms/android/nimbusjs/src/androidTest/java/NimbusJSUtilsTests.kt]: nit: Preferably we shouldn't use keywords as variable names.

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import android.webkit.WebView
+import java.io.*
+import java.nio.charset.StandardCharsets
+
+class NimbusJSUtilities() {
+    companion object Injection {
+
+        fun injectedNimbusStream(inputStream: InputStream, context: Context): InputStream {
+            val jsString = ResourceUtils(context).stringFromRawResource(R.raw.nimbus)
+            var html: String? = this.readAssetStream(inputStream)
```
Comment (jamie-houston) [inline platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: looks like this shouldn't have a null html... so maybe
```
var html = this.readAssetStream(inputStream) ?: throw MyAwesomeException()
```

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            if (html!!.contains("<head>")) {
+                html = html.replace("<head>", "<head>\n$jsString\n")
+            } else if (html.contains("</head>")) {
```
Comment (jamie-houston) [inline platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: How does the html end up with either \<head\> or \</head\>?  Seems weird that it'd be missing the first but have the second...?

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            }
+            return ByteArrayInputStream(html!!.toByteArray(StandardCharsets.UTF_8))
+        }
+
+        private fun readAssetStream(stream: InputStream): String? {
+            try {
+                val bufferSize = 1024
+                val buffer = CharArray(bufferSize)
+                val out = StringBuilder()
+                val `in`: Reader = InputStreamReader(stream, "UTF-8")
```
Comment (jamie-houston) [inline platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: nit: Use constant here too (`StandardCharsets.UTF_8`)

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt (hunk 3 above)
Comment (paultiarks) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: In cases where `<head>` has attributes. We could probably shorten this to just checking for `</head>`

File: platforms/android/nimbusjs/src/androidTest/java/NimbusJSUtilsTests.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    @Before
+    fun setup() {
+        context = InstrumentationRegistry.getInstrumentation().targetContext
+    }
+
+    @Test
+    fun testInjectionTwo() {
+        var inputStream: InputStream = context.assets.open("testPage.html")
+        inputStream = NimbusJSUtilities.injectedNimbusStream(inputStream, context)
+        var resultString = readAssetStream(inputStream)
```
Comment (jamie-houston) [inline platforms/android/nimbusjs/src/androidTest/java/NimbusJSUtilsTests.kt]: I think you can just use the helper `inputStream.bufferedReader().readText()` or something like that.

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt (hunk 2 above)
Comment (jamie-houston) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: here too, I think you could just read it in with `inputStream.bufferedReader(StandardCharsets.UTF_8).readText()` and not create the (duplicated) helper method

Comment (jamie-houston) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: There's probably a way to filter it while you're reading it too, so if you wanted to just search for \<head\> and modify or copy to new string/stream you could.  But not sure the (perf) benefit of the added complexity.

File: platforms/android/nimbusjs/src/androidTest/java/NimbusJSUtilsTests.kt (hunk 5 above)
Comment (paultiarks) [reply platforms/android/nimbusjs/src/androidTest/java/NimbusJSUtilsTests.kt]: Wow that cleans things up considerably. Thank you

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            } catch (e: Exception) {
+                Log.e("Nimbus", "Unable to inject Nimbus", e)
+            }
+            return ""
+        }
+    }
+}
+
+private fun String.Companion.fromStream(stream: InputStream): String {
+    val result = ByteArrayOutputStream()
```
Comment (jamie-houston) [inline platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: Is this doing something different than the other stream to string functions?  Seems like the same thing...?

Also, I'm not sure of this paradigm... creating an extension method to String that is only used in this one method.  If you need this, I'd say you could make it a private function inside of the ResourceUtils class.

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt (hunk 3 above)
Comment (jamie-houston) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: ahh... yeah that makes sense!

Comment (jamie-houston) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: Do you want to handle a case where there is no head section?  Not sure if that's valid html, but maybe you'd add the \<head\> if it was missing...? 🤷‍♂

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt (hunk 6 above)
Comment (paultiarks) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: yep. good call, cleaning up now.

File: platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt (hunk 3 above)
Comment (paultiarks) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: I'm honestly not sure what to do in that case. I would lean away from adding a `<head>` if it's missing entirely. But I could be convinced.

Comment (jamie-houston) [reply platforms/android/nimbusjs/src/main/java/com/salesforce/nimbusjs/NimbusJSUtilities.kt]: /shrug

File: platforms/android/nimbusjs/buildNimbusJS.sh [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+cd ../../packages/nimbus-bridge/
+
+npm install
+
+npm run build
+
+mkdir ../../platforms/android/nimbusjs/src/main/res/raw/
+
+cp dist/iife/nimbus.js ../../platforms/android/nimbusjs/src/main/res/raw/nimbus.js
```
Comment (wes566) [inline platforms/android/nimbusjs/buildNimbusJS.sh]: the android browser supports es modules so we should be using the ESM version of nimbus instead of iife, unless I'm missing something

Comment (paultiarks) [reply platforms/android/nimbusjs/buildNimbusJS.sh]: If I remember correctly, there are older android versions that don't yet support it, so I used the iife version
